core/ssl-only/class-edd-slack-app-oauth-settings.php

```diff
@@ -269,148 +319,214 @@ public function add_slack_invites_oauth_register( $args ) {
+
+		$redirect_uri = add_query_arg(
+			array(
+				'post_type' => 'download',
+				'page'      => 'edd-settings',
+				'tab'       => 'extensions',
+				'section'   => 'edd-slack-settings',
+			),
+			admin_url( 'edit.php' )
+		);
+
 		// If we need to get an OAUTH Token
-		// $_GET['state'] is set by the JavaScript for the OAUTH2 Popup
-		// $_GET['section'] is set properly by our redirect_uri
-		if ( isset( $_GET['code'] ) && 
-			isset( $_GET['state'] ) && 
-			$_GET['state'] == 'saving' && 
-			isset( $_GET['token_type'] ) &&
-			isset( $_GET['section'] ) && 
-			$_GET['section'] == 'edd-slack-settings' && 
-			( ( ! $oauth_token || $oauth_token == '-1' ) && $_GET['token_type'] == 'main' || 
-			 ! edd_get_option( 'slack_app_has_client_scope' ) && $_GET['token_type'] == 'team_invites' ) ) {
-			
-			$client_id = edd_get_option( 'slack_app_client_id' );
+		if ( ( ! $this->is_authorized() && 'main' === $token_type ) || ( ! edd_get_option( 'slack_app_has_client_scope' ) && 'team_invites' === $token_type ) ) {
+
+			$client_id     = edd_get_option( 'slack_app_client_id' );
 			$client_secret = edd_get_option( 'slack_app_client_secret' );
-		
-			$redirect_uri = urlencode_deep( admin_url( 'edit.php?post_type=download&page=edd-settings&tab=extensions&section=edd-slack-settings' ) );
-			
-			$oauth_access_url = add_query_arg(
+			$oauth_request = EDDSLACK()->slack_api->post(
+				'oauth.access',
 				array(
-					'client_id' => $client_id,
-					'client_secret' => $client_secret,
-					'code' => $_GET['code'],
-					'redirect_uri' => $redirect_uri,
-				),
-				'oauth.access'
-			);
-			
-			$oauth_request = EDDSLACK()->slack_api->post( 
-				$oauth_access_url
+					'body' => array(
+						'client_id'     => $client_id,
+						'client_secret' => $client_secret,
+						'code'          => $code,
+						'redirect_uri'  => $redirect_uri,
+					),
+				)
 			);
-			
-			if ( $oauth_request->ok == 'true' ) {
-				
-				$oauth_token = $oauth_request->access_token;
-				EDDSLACK()->slack_api->set_oauth_token( $oauth_token );
-				
-				if ( $_GET['token_type'] == 'main' ) {
-				
-					$this->admin_notices[] = array(
-						'edd-notices',
-						'edd_slack_app_auth',
-						_x( 'Slack App Linked Successfully.', 'EDD Slack App Auth Successful', 'edd-slack' ),
-						'updated'
-					);
-					
-				}
-				else if ( $_GET['token_type'] == 'team_invites' ) {
-					
-					$granted_client_scope = edd_update_option( 'slack_app_has_client_scope', true );
-					
-					$this->admin_notices[] = array(
-						'edd-notices',
-						'edd_slack_app_team_invites_auth',
-						_x( 'Slack App Team Invites Enabled Successfully.', 'EDD Slack App Team Invites Auth Successful', 'edd-slack' ),
-						'updated'
-					);
-					
-				}
-				
+
+			if ( empty( $oauth_request->ok ) ) {
+				$redirect_uri = add_query_arg(
+					array(
+						'edd_slack_oauth' => 'fail',
+						'edd_slack_type'  => 'error',
+					),
+					$redirect_uri
+				);
+
+				wp_safe_redirect( $redirect_uri );
```

Comment:
```suggestion
				wp_safe_redirect( $redirect_uri );
				exit;
```

Without the exit, it actually keeps executing the rest of the code and you end up on the success page.